[Standard library] Ascribe types to some literals to improve typecheck time

DougGregor opened the pull request:
One expression in the new hashing implementation is going exponential,
accounting for a huge amount of type-checking type. Add (admittedly ugly)
“as UInt64” annotations to greatly reduce the time to type-check this
expression.

Do the same for some of the UnicodeScalar decoding code.

*Ahem* type-checking time for the standard library goes from 24s->12s with
this change. Added a type-checker “slow” performance test and captured
the problem in rdar://problem/42672946.

DougGregor committed:
[Type checker] Eliminate some unnecessary TypeChecker uses. NFC

DougGregor committed:
[Standard library] Ascribe types to some literals to improve typecheck time.

One expression in the new hashing implementation is going exponential,
accounting for a huge amount of type-checking type. Add (admittedly ugly)
“as UInt64” annotations to greatly reduce the time to type-check this
expression.

*Ahem* type-checking time for the standard library goes from 24s->14s with
this change. Added a type-checker “slow” performance test and captured
the problem in rdar://problem/42672946.

DougGregor commented:
@lorentey might be interested in this :)

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test

jckarter commented on stdlib/public/core/Hasher.swift:
```diff
@@ -90,7 +90,10 @@ internal struct _HasherTailBuffer {
+    // FIXME: The "as UInt64" annotations here eliminate some exponential
+    // behavior in the expression type checker <rdar://problem/42672946>.
+    _sanityCheck(tail & ~((1 as UInt64) << ((byteCount & (7 as UInt64))
+          << (3 as UInt64)) - (1 as UInt64)) == (0 as UInt64))
```
`as` still has to consider bridging as a possibility. Is there any more blood to be squeezed from this stone by assigning these to explicitly-typed constants instead?

```
let one: UInt64 = 1
let seven: UInt64 = 7
let three: UInt64 = 3
tail & ~(one << ...
```

DougGregor replied:
Maybe! I'm not sure it'll make enough difference to matter, but maybe it'll be less ugly.

xedin replied:
Another possibility after SE-0213 is to do `UInt64(1)` since that's going to be an ascription as well for literals.

DougGregor replied:
I found a formulation that's faster to type-check.

DougGregor committed:
[Standard library] Faster formulation for the _HasherTailBuffer sanity check

DougGregor committed:
[Standard library] Break up another “large” expression.

Bitwise arithmetic is a killer for the type checker :(. Shave off another
half second of typecheck time in the standard library.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test
